dags/xcom_gcs_ds_k8sExecutor.py

```diff
@@ -0,0 +1,220 @@
+                )
+            }
+
+
+@dag(
+    start_date=datetime(2021, 1, 1),
+    schedule_interval=None,
+    catchup=False,
+    doc_md=docs
+)
+def using_gcs_for_xcom_ds_k8sExec():
+
+    @task(executor_config=load_config)
+    def load_data():
+        """Pull Census data from Public BigQuery and save as Pandas dataframe in GCS bucket with XCom"""
+
+        bq = BigQueryHook()
+        sql = """
+        SELECT * FROM `bigquery-public-data.ml_datasets.census_adult_income`
+        """
+
+        return bq.get_pandas_df(sql=sql, dialect='standard')
+
+
+
+    @task
+    def preprocessing(df: pd.DataFrame):
+        """Clean Data and prepare for feature engineering
+        
+        Returns pandas dataframe via XCom to GCS bucket.
+
+        Keyword arguments:
+        df -- Raw data pulled from BigQuery to be processed. 
+        """
+
+        df.dropna(inplace=True)
+        df.drop_duplicates(inplace=True)
+
+        # Clean Categorical Variables (strings)
+        cols = df.columns
+        for col in cols:
+            if df.dtypes[col]=='object':
+                df[col] =df[col].apply(lambda x: x.rstrip().lstrip())
+
+
+        # Rename up '?' values as 'Unknown'
+        df['workclass'] = df['workclass'].apply(lambda x: 'Unknown' if x == '?' else x)
+        df['occupation'] = df['occupation'].apply(lambda x: 'Unknown' if x == '?' else x)
+        df['native_country'] = df['native_country'].apply(lambda x: 'Unknown' if x == '?' else x)
+
+
+        # Drop Extra/Unused Columns
+        df.drop(columns=['education_num', 'relationship', 'functional_weight'], inplace=True)
+
+        return df
+
+    @task
+    def feature_engineering(df: pd.DataFrame):
+        """Feature engineering step
+        
+        Returns pandas dataframe via XCom to GCS bucket.
+
+        Keyword arguments:
+        df -- data from previous step pulled from BigQuery to be processed. 
+        """
+
+        
+        # Onehot encoding 
+        df = pd.get_dummies(df, prefix='workclass', columns=['workclass'])
+        df = pd.get_dummies(df, prefix='education', columns=['education'])
+        df = pd.get_dummies(df, prefix='occupation', columns=['occupation'])
+        df = pd.get_dummies(df, prefix='race', columns=['race'])
+        df = pd.get_dummies(df, prefix='sex', columns=['sex'])
+        df = pd.get_dummies(df, prefix='income_bracket', columns=['income_bracket'])
+        df = pd.get_dummies(df, prefix='native_country', columns=['native_country'])
+
+
+        # Bin Ages
+        df['age_bins'] = pd.cut(x=df['age'], bins=[16,29,39,49,59,100], labels=[1, 2, 3, 4, 5])
+
+
+        # Dependent Variable
+        df['never_married'] = df['marital_status'].apply(lambda x: 1 if x == 'Never-married' else 0) 
+
+
+        # Drop redundant colulmn
+        df.drop(columns=['income_bracket_<=50K', 'marital_status', 'age'], inplace=True)
+
+        return df
+
+
+    @task(executor_config=modeling_config)
+    def train(df: pd.DataFrame):
```

Comment:
Would it make more sense to call this function something like `validate` since it's not so much returning a trained model as a metric that is used to decide whether or not to use the model? I could see an extension of this step where versions of the model are checkpointed and the best iteration is saved/passed instead.

Reply:
Yes that would be a great add on. Do we want to make this change on the first example dag as well? We could even make it more descriptive as well like "cross_validation" if that make sense.

Reply:
Yup, that makes sense.